internal/session/sessionwantsender.go

```diff
@@ -354,21 +361,59 @@ func (spm *sessionWantSender) processUpdates(updates []update) {
 		}
 	}
 
-	// If all available peers for a cid sent a DONT_HAVE, signal to the session
-	// that we've exhausted available peers
-	if dontHaves.Len() > 0 {
-		exhausted := spm.bpm.AllPeersDoNotHaveBlock(spm.peerAvlMgr.availablePeers(), dontHaves.Keys())
-		newlyExhausted := spm.newlyExhausted(exhausted)
-		if len(newlyExhausted) > 0 {
-			spm.onPeersExhausted(newlyExhausted)
-		}
-	}
-
 	// If any peers have sent us too many consecutive DONT_HAVEs, remove them
 	// from the session
 	for p := range prunePeers {
 		spm.SignalAvailability(p, false)
 	}
+
+	return dontHaves
+}
+
+// checkForExhaustedWants checks if there are any wants for which all peers
+// have sent a DONT_HAVE. We call these "exhausted" wants.
+func (spm *sessionWantSender) checkForExhaustedWants(dontHaves *cid.Set, newlyUnavailable []peer.ID) {
+	// If there are no new DONT_HAVEs, and no peers became unavailable, then
+	// we don't need to check for exhausted wants
+	if dontHaves.Len() == 0 && len(newlyUnavailable) == 0 {
+		return
+	}
+
+	// We need to check each want for which we just received a DONT_HAVE
+	wants := dontHaves.Keys()
+
+	// If a peer just became unavailable, then we need to check all wants
+	// (because it may be the last peer who hadn't sent a DONT_HAVE for a CID)
+	if len(newlyUnavailable) > 0 {
+		// Collect all pending wants
+		wants = make([]cid.Cid, len(spm.wants))
+		for c := range spm.wants {
+			wants = append(wants, c)
+		}
+
+		// If the last available peer in the session has become unavailable
+		// then we need to broadcast all pending wants
+		if len(spm.peerAvlMgr.availablePeers()) == 0 {
+			spm.processExhaustedWants(wants)
+			return
+		}
+	}
+
+	// If all available peers for a cid sent a DONT_HAVE, signal to the session
+	// that we've exhausted available peers
+	if len(wants) > 0 {
```

Comment:
All wants from all peers if `newlyUnavailable > 0`?

Reply:
Let me know if the explanation above makes sense, I think it covers this too.

Reply:
Ah, I missed the fact that we're _filtering_ this list. This makes sense.